kv/kvserver: TestReplicateQueueSeesLearnerOrJointConfig failed

kv/kvserver.TestReplicateQueueSeesLearnerOrJointConfig [failed](https://tanzanite.cluster.engflow.com/invocations/default/cb81f97d-ba0c-4168-a132-161a00cf736b?testReportRun=3&testReportShard=30&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on master @ [8c5af013aa725d4f441abeff282894ff22f04ba7](https://github.com/cockroachdb/cockroach/commits/8c5af013aa725d4f441abeff282894ff22f04ba7):


```
=== RUN   TestReplicateQueueSeesLearnerOrJointConfig
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestReplicateQueueSeesLearnerOrJointConfig261065314
    test_log_scope.go:76: use -show-logs to present logs inline
    replica_learner_test.go:1276: condition failed to evaluate within 45s: from replica_learner_test.go:1279: Mismatch in num voters [(n1,s1):1 (n2,s2):3], desc: r75:/{Table/Max-Max} [(n1,s1):1, (n2,s2):3, next=5, gen=10, sticky=9223372036.854775807,2147483647]
    panic.go:629: -- test log scope end --
test logs left over in: outputs.zip/logTestReplicateQueueSeesLearnerOrJointConfig261065314
--- FAIL: TestReplicateQueueSeesLearnerOrJointConfig (51.31s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=3</code>
 - <code>shard=30</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicateQueueSeesLearnerOrJointConfig.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-48211

--- a/pkg/kv/kvserver/replica_learner_test.go
+++ b/pkg/kv/kvserver/replica_learner_test.go
@@ -1216,9 +1216,23 @@ func TestReplicateQueueSeesLearnerOrJointConfig(t *testing.T) {
 	// NB also see TestAllocatorRemoveLearner for a lower-level test.
 
 	ctx := context.Background()
-	knobs, ltk := makeReplicationTestKnobs()
+	_, ltk := makeReplicationTestKnobs()
+	// targetRangeID is the ID of the only range for which we will enable the
+	// replicate queue. This is the range the test makes assertions for. If other
+	// ranges get enqueued as well, it might take longer to process the range the
+	// test cares about. This is particularly important for leader leases because
+	// the target range is the only one with RF=3; other ranges have RF=1 and thus
+	// get leaders and leaseholder much faster as they don't need to wait for
+	// store liveness support.
+	var targetRangeID atomic.Value
+	ltk.storeKnobs.BaseQueueDisabledBypassFilter = func(rangeID roachpb.RangeID) bool {
+		if target := targetRangeID.Load(); target != nil && rangeID == target {
+			return true
+		}
+		return false
+	}
 	tc := testcluster.StartTestCluster(t, 3, base.TestClusterArgs{
-		ServerArgs:      base.TestServerArgs{Knobs: knobs},
+		ServerArgs:      base.TestServerArgs{Knobs: base.TestingKnobs{Store: &ltk.storeKnobs}},
 		ReplicationMode: base.ReplicationManual,
 	})
 	defer tc.Stopper().Stop(ctx)
@@ -1230,11 +1244,12 @@ func TestReplicateQueueSeesLearnerOrJointConfig(t *testing.T) {
 		_ = tc.AddVotersOrFatal(t, scratchStartKey, tc.Target(1))
 	})
 
-	// Run the replicate queue.
+	// Run the replicate queue only for our target range.
 	store, repl := getFirstStoreReplica(t, tc.Server(0), scratchStartKey)
 	{
+		// Set our target range ID in the knobs.
+		targetRangeID.Store(repl.RangeID)
 		require.Equal(t, int64(0), getFirstStoreMetric(t, tc.Server(0), `queue.replicate.removelearnerreplica`))
-		store.TestingSetReplicateQueueActive(true)
 		traceCtx, finish := tracing.ContextWithRecordingSpan(ctx, store.GetStoreConfig().Tracer(), "trace-enqueue")
 		processErr, err := store.Enqueue(
 			traceCtx, "replicate", repl, true /* skipShouldQueue */, false, /* async */
@@ -1255,15 +1270,16 @@ func TestReplicateQueueSeesLearnerOrJointConfig(t *testing.T) {
 			}
 			return nil
 		})
-		// It has done everything it needs to do now, disable before the next test section.
-		store.TestingSetReplicateQueueActive(false)
+		// Unset the target range ID before the next test.
+		targetRangeID.Store(roachpb.RangeID(0))
 	}
 
 	// Create a VOTER_OUTGOING, i.e. a joint configuration.
 	ltk.withStopAfterJointConfig(func() {
+		// Set our target range ID in the knobs.
+		targetRangeID.Store(repl.RangeID)
 		desc := tc.RemoveVotersOrFatal(t, scratchStartKey, tc.Target(2))
 		require.True(t, desc.Replicas().InAtomicReplicationChange(), desc)
-		store.TestingSetReplicateQueueActive(true)
 		traceCtx, finish := tracing.ContextWithRecordingSpan(ctx, store.GetStoreConfig().Tracer(), "trace-enqueue")
 		processErr, err := store.Enqueue(
 			traceCtx, "replicate", repl, true /* skipShouldQueue */, false, /* async */
@@ -1280,7 +1296,7 @@ func TestReplicateQueueSeesLearnerOrJointConfig(t *testing.T) {
 			}
 			return nil
 		})
-		store.TestingSetReplicateQueueActive(false)
+		targetRangeID.Store(roachpb.RangeID(0))
 	})
 }
 

